core/src/main/java/org/apache/druid/math/expr/Expr.java

```diff
@@ -171,6 +174,12 @@ default boolean canVectorize(InputBindingInspector inspector)
     throw Exprs.cannotVectorize(this);
   }
 
+  @Override
+  default byte[] getCacheKey()
+  {
+    return new CacheKeyBuilder(EXPR_CACHE_KEY).appendString(stringify()).build();
```

Comment:
`stringify` is somewhat expensive and this will be called once per segment. Can we cache it, like we cache the original and parsed expressions? Best way I can think of right now is to implement that caching in ExpressionFilter, ExpressionVirtualColumn, etc.

Btw, that caching should be lazy, for two reasons:

- otherwise we'll waste time computing it when it'll never be needed
- sometimes the cache key will be incomputable (lookup doesn't exist, or error retrieving it) and we don't want that to totally prevent instantiating the VirtualColumn or DimFilter objects

Reply:
added caching similar to the `Expr` themselves